Upsert optimizations

wesleykendall opened the pull request:
@wesokes @jaredlewis  this adds the following optimizations and capabilities to the newer upsert method:

1. Previously every upsert would update every row in the upsert even if no data had changed. Now upsert defaults to only updating a row if the row has changed. I used the solution at https://stackoverflow.com/questions/41482441/postgresql-upsert-do-nothing-if-fields-dont-change for this optimization
2. Because of 1), we now have the ability to return "untouched" rows during an upsert. I.e. we have a fourth category (created, updated, untouched, and deleted rows)

The optimization for not updating duplicate rows has some pretty big advantages:

1. The row isn't locked for an update, leaving less room for deadlocks
2. Because of postgres's MVCC stuff, it would previously create a new entry on disk for this updated row along with updating the index. It avoids this penalty now.
3. Triggers for updates are never executed

This optimization required having to do some special stuff to get the untouched rows though. When ignoring duplicate updates, the results are not returned by the `RETURNING` clause. I used the solution from the second answer to https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql do this. It creates a CTE for the inserted values and then merges the results with that to figure out which ones weren't touched in the upsert operation.


## Changes to interface
Because of these additions, there are two new options to bulk upsert and some interface changes:

1. `bulk_upsert2` and `sync2` return a list of results rather than tuples. It's a wrapper class that I wrote that allows one to do `results.created`, `results.untouched`, `results.updated`, and `results.deleted`. This avoids having to do stuff like `created, inserted, untouched = upsert2`. And it's more compatible with the original interface. 
2. `ignore_duplicate_updates` is an arg to `upsert2` and `sync2` now. By default, we set this to True. I believe this would be the expected behavior during a bulk upsert. If this is disabled and you provide fields to updates, there will never be untouched rows in the result. They will always be in the `updated` section. The only use case I can think of this is if someone wants triggers to fire or other things during the sync.
3. `return_untouched` is an arg to `upsert2`. It defaults to `False`. Returning the untouched rows of a bulk upsert operation has more overhead than just a native upsert because of the temporary table and join. This argument is always `True` when doing a sync since the sync operation needs to know every row that was touched in order to delete the ones that weren't.

wesokes commented:
@wesleykendall looks good to me except for a few formatting things. when args go to multiple lines on function definitions / calls, we keep everything at a single level of indention like
```
def func(
    one, two, ...,
    more, here, ...,
    and, here, ...
):
```
rather than indenting the all lines to the first line's first arg.

wesleykendall commented:
Thanks @wesokes I addressed comments in https://github.com/ambitioninc/django-manager-utils/pull/104/commits/29e324b7f848c5c9e61d284d5582154e079e4de5. Will merge and deploy once tests pass
